### PR TITLE
Move html gen to build.d

### PR DESCRIPTION
**Pull request by marler8997**

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -533,6 +534,55 @@ alias zip = memoize!(function()
...
+    if (env.get("DMD", null).length == 0) {
+        htmlDep.commandFunction = () {
+            writefln("ERROR: cannot build '%s' unless DMD is set", htmlDep.name);
+        };
+    } else {
+        htmlDep.msg = "HTML " ~ docSources.join(" ");
+        htmlDep.deps = new DependencyRef[docSources.length];
+        const stddocs = env.get("STDDOC", "").split();
+        foreach(i; 0 .. docSources.length) {
+            Dependency docDep;
```
Comment (MoonlightSentinel) [inline src/build.d]: Maybe split this off into another local function and declare the dependencies via an UFCS chain?

File: src/build.d [hunk 2]
```diff
@@ -533,6 +534,55 @@ alias zip = memoize!(function()
...
+                env["MODEL_FLAG"],
+                "-c",
+                "-Dd" ~ env["DOCSRC"],
+                "-J"~env["G"], "-J../res",
+                "-I" ~ env["D"],
+                srcDir.buildPath("project.ddoc")
+            ] ~ stddocs ~ [
+                "-Df" ~ docDep.target,
+                docSources[i]
+            ] ~ flags["DFLAGS"];
```
Comment (MoonlightSentinel) [inline src/build.d]: `env["MODEL_FLAG"]` and, `"-J"~env["G"]` are contained in `DFLAGS`

File: src/posix.mak [hunk 3]
```diff
@@ -361,7 +333,7 @@ SRC_MAKE = posix.mak osmodel.mak
 
 DEPS = $(patsubst %.o,%.deps,$(DMD_OBJS))
 
-RUN_BUILD = $(GENERATED)/build HOST_DMD="$(HOST_DMD)" CXX="$(HOST_CXX)" OS=$(OS) BUILD=$(BUILD) MODEL=$(MODEL) AUTO_BOOTSTRAP="$(AUTO_BOOTSTRAP)" --called-from-make
+RUN_BUILD = $(GENERATED)/build HOST_DMD="$(HOST_DMD)" CXX="$(HOST_CXX)" OS=$(OS) BUILD=$(BUILD) MODEL=$(MODEL) AUTO_BOOTSTRAP="$(AUTO_BOOTSTRAP)" DOCDIR="$(DOCDIR)" STDDOC="$(STDDOC)" DOC_OUTPUT_DIR="$(DOC_OUTPUT_DIR)" DMD="$(DMD)" --called-from-make
```
Comment (MoonlightSentinel) [inline src/posix.mak]:
```suggestion
RUN_BUILD = $(GENERATED)/build HOST_DMD="$(HOST_DMD)" CXX="$(HOST_CXX)" OS=$(OS) BUILD=$(BUILD) MODEL=$(MODEL) AUTO_BOOTSTRAP="$(AUTO_BOOTSTRAP)" DOCDIR="$(DOCDIR)" STDDOC="$(STDDOC)" DOC_OUTPUT_DIR="$(DOC_OUTPUT_DIR)" --called-from-make
```

File: src/build.d [hunk 4]
```diff
@@ -533,6 +534,55 @@ alias zip = memoize!(function()
...
+            docDep.msg = docDep.command.join(" ");
```
Comment (MoonlightSentinel) [inline src/build.d]:
```suggestion
            docDep.msg = docDep.target
```

Less noisy, `--dry-run` already exists to dump the invoked command

File: src/build.d [hunk 5]
```diff
@@ -533,6 +534,55 @@ alias zip = memoize!(function()
...
+    Dependency htmlDep;
+    htmlDep.name = "html";
+    htmlDep.description = "Generate html docs, requires DMD and STDDOC to be set";
+    auto docSources = .sources.root ~ .sources.lexer ~ .sources.dmd ~ env["D"].buildPath("frontend.d");
+    if (env.get("DMD", null).length == 0) {
+        htmlDep.commandFunction = () {
+            writefln("ERROR: cannot build '%s' unless DMD is set", htmlDep.name);
+        };
+    } else {
+        htmlDep.msg = "HTML " ~ docSources.join(" ");
```
Comment (MoonlightSentinel) [inline src/build.d]:
```suggestion
        htmlDep.msg = "HTML";
```
The filenames are already printed from the dependencies

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]
Some suggestions

File: src/posix.mak [hunk 6]
```diff
@@ -312,17 +312,6 @@ endif
 
 ######## DMD frontend source files
 
-FRONT_SRCS=$(addsuffix .d, $(addprefix $D/,access aggregate aliasthis apply argtypes argtypes_sysv_x64 arrayop	\
```
Comment (atilaneves) [inline src/posix.mak]: Are all these deletions actually related to html generation?

Comment (atilaneves) [reply src/posix.mak]: And then users won't be able to use the makefile for that anymore?

**Comment by wilzbach** [comment]
Mind the DAutotTest logs:

```diff
diff --git a/web/phobos-prerelease/dmd_access.html b/web/phobos-prerelease/dmd_access.html
index 5315855..957b499 100644
--- a/web/phobos-prerelease/dmd_access.html
+++ b/web/phobos-prerelease/dmd_access.html
@@ -162,7 +162,7 @@
 		<div >			If you spot a problem with this page, click here to create a Bugzilla issue.
 		</div>
 	</div>
-	<div class="tip smallprint">		<a href="https://github.com/dlang/dmd/edit/master/dmd/access.d">Improve this page</a>
+	<div class="tip smallprint">		<a href="https://github.com/dlang/dmd/edit/master//dev/shm/dtest/work/repo/dmd/src/dmd/access.d">Improve this page</a>
 		<div >			Quickly fork, edit online, and submit a pull request for this page.
 			Requires a signed-in GitHub account. This works well for small changes.
 			If you'd like to make larger changes you may want to consider using
```

File: src/build.d [hunk 7]
```diff
@@ -533,6 +534,55 @@ alias zip = memoize!(function()
...
+    return htmlFilePrefix ~ ".html";
+}
+
+alias html = memoize!(function()
+{
+    Dependency htmlDep;
+    htmlDep.name = "html";
+    htmlDep.description = "Generate html docs, requires DMD and STDDOC to be set";
+    auto docSources = .sources.root ~ .sources.lexer ~ .sources.dmd ~ env["D"].buildPath("frontend.d");
+    if (env.get("DMD", null).length == 0) {
```
[the remark this answers is not in the document]
Comment (marler8997) [reply src/build.d]: The documentation builder uses the `DMD` environment variable to select the compiler to generate the documentation.

File: src/build.d (hunk 1 above)
Comment (marler8997) [reply src/build.d]: What's nice about the current implementation is that I can create slices into the docSources array so I don't have to allocate a new array for each dependency.  Using a UFCS chain would make that difficult.

File: src/build.d (hunk 4 above)
Comment (marler8997) [reply src/build.d]: we want to see the commands in the CI builders though.`--dry-run` is for developers

File: src/posix.mak (hunk 3 above)
Comment (marler8997) [reply src/posix.mak]: doc CI builder is using the `DMD` variable to pass in the compiler to use for doc generation

File: src/build.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/build.d]: Slicing instead of reallocating is indeed a good idea. You could keep it by using `docSources.slide(1)`

**Comment by marler8997** [comment]
> Mind the DAutotTest logs: ...

I switched to passing relative paths for the main source file in ddoc generation and this seems to have fixed the ddoc links.

**Comment by marler8997** [comment]
Getting this error on the GDC build...is `slide` a new function or was it moved at some point?
```
build.d:555:35: error: function std.range.slide!(cast(Flag)true, string[]).slide is not accessible from module build
  555 |         docDeps = docSources.slide(1).map!(sourceArray => {
      |      
```

**Comment by marler8997** [comment]
Actually it looks like `chunks` should work as well, and we'are already using that one so it should work.

**Comment by wilzbach** [comment]
> Getting this error on the GDC build...is slide a new function or was it moved at some point?

Yeah, I added it in 2.079.

https://github.com/dlang/phobos/pull/5943
https://dlang.org/changelog/2.079.0.html#std-range-slide

GDC 8/9 are based on Phobos 2.076.

**Review by PetarKirov — APPROVED** [review]

**Pull request merged**